wip: markets page recommended tab

gomesalexandre opened the pull request:
## Description

<!-- Please describe your changes -->

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes #

## Risk

> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)

<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->

gomesalexandre commented on packages/utils/src/evmChainIds.ts:
```diff
@@ -0,0 +1,20 @@
+import type { EvmChainId } from '@shapeshiftoss/types'
+import { KnownChainIds } from '@shapeshiftoss/types'
+
+export const isEvmChainId = (
+  maybeEvmChainId: string | EvmChainId,
+): maybeEvmChainId is EvmChainId => {
+  return evmChainIds.includes(maybeEvmChainId as EvmChainId)
+}
+
+export const evmChainIds = [
+  KnownChainIds.EthereumMainnet,
+  KnownChainIds.AvalancheMainnet,
+  KnownChainIds.OptimismMainnet,
+  KnownChainIds.BnbSmartChainMainnet,
+  KnownChainIds.PolygonMainnet,
+  KnownChainIds.GnosisMainnet,
+  KnownChainIds.ArbitrumMainnet,
+  KnownChainIds.ArbitrumNovaMainnet,
+  KnownChainIds.BaseMainnet,
+] as const
```
Yes, this is duplicated but fixes this guy when running `tsx ./scripts/generateAssetData/generateAssetData.ts ` following portals utils consolidation in web:

https://github.com/ethers-io/ethers.js/discussions/4163

We already have it fixed in web through react-app-rewired config but trying to get this to work in scripts is borderline impossible

woodenfurniture replied:
Actually moved it directly into the script in #7744 so as not to introduce 2 sources of truth to the rest of the project

woodenfurniture replied:
i.e can remove this after rebase

woodenfurniture replied:
Oh i see this PR is after that change. Ignore me.

gomesalexandre commented on src/pages/Buy/components/Sparkline.tsx:
```diff
@@ -28,6 +28,7 @@ export const SparkLine: React.FC<SparkLineProps> = ({
   themeColor = 'blue',
   height = 50,
 }) => {
+  console.log({ assetId, percentChange })
```
rm

gomesalexandre commented:
Opening for early review - still some cleanup to do here and oil to be bearded, this looks jank visually atm

gomesalexandre commented on packages/utils/src/index.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/arbitrum/index.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/avalanche/index.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/base/index.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/bnbsmartchain/index.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/coingecko.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/cosmos/index.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/ethTokens/uniswap.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/ethereum/index.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/generateColorMap.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/gnosis/index.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/polygon/index.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on scripts/generateAssetData/utils/portals.ts:
👋🏽 now all living in web

gomesalexandre commented on src/lib/asset-service/service/color-map.json:
Now living in web too 👋🏽

gomesalexandre commented on src/lib/market-service/portals/portals.ts:
Plumbing of Portals utils consolidation in web

gomesalexandre commented on src/lib/portals/utils.ts:
```diff
@@ -1,17 +1,266 @@
-} from 'lib/market-service/portals/types'
+} from './types'
+
+// Accommodate for script vs. web shenanigans
```
i.e `getConfig()` *would* technically be available, but since web won't be running, won't work

woodenfurniture replied:
[question] Are these utils not intended for the client, wouldnt it be better to keep them separate to the client code? That way we arent adding code that is known not to work here

gomesalexandre replied:
Just to disambiguate things, this *is* client code atm, and moving it out of `src/lib` would still be client code (except if we move portals, and consequently asset gen to packages, which is absolute hell and we shouldn't do, spent a day trying to doso 💀). Client meaning within the web tsc project and not its separate package.

It is intended to (and is) consumed in both web running (as in, *running* web build) and in the context of scripts, (as in, web *project*, but web itself not running) meaning `getConfig()` isn't guaranteed to work. It should really, but...

Noticed two missing occurences of `getConfig()` however: 59d8d6586c

@woodenfurniture if this is confusing, happy to remove the paranoia short circuit, we really shouldn't need it!

woodenfurniture replied:
yes if we dont need it we should totally remove it

gomesalexandre commented on src/pages/Markets/Recommended.tsx:
```diff
@@ -1,18 +1,274 @@
+      {
+        title: translate('markets.categories.topMovers.title'),
+        component: (
+          <AssetsGrid
```
add TODO here too

gomesalexandre replied:
cbce36f365

gomesalexandre commented on src/pages/Markets/hooks/usePortalsAssetsQuery.ts:
```diff
@@ -0,0 +1,74 @@
+      ? () =>
+          fetchPortalsTokens({
+            limit: '10',
+            chainIds: undefined,
```
Handle actual selected ChainId here

gomesalexandre commented on src/pages/Markets/hooks/usePortalsAssetsQuery.ts:
```diff
@@ -0,0 +1,74 @@
+
+          return {
+            asset,
+            // TODO(gomes): do we even need TokenInfo here? Market-data should contain all we need and we shouldn't need to use raw Portals data?
```
We probably don't, now that we upsert

gomesalexandre commented on src/lib/market-service/portals/portals.ts:
```diff
@@ -115,11 +113,7 @@ export class PortalsMarketService implements MarketService {
-    const assets = generatedAssetData as unknown as AssetsByIdPartial
-
     try {
-      const asset = assets[assetId]
-      if (!asset) return null
```
That was a big rug - this could never work with runtime assets, since these do *not* exist in generatedAssetData.json by definition

woodenfurniture commented on src/pages/Markets/Recommended.tsx:
```diff
@@ -1,18 +1,281 @@
+          <LpGrid
+            assetIds={assetIds}
+            selectedChainId={selectedChainId}
+            // TODO(gomes): loading state when implemented
```
[question] do we want these implemented in this PR or in follow-up?

gomesalexandre replied:
That's on purpose! Loading state *is* implemented, but all are leveraging Portals loading state as a monkey patch currently until others are implemented

woodenfurniture commented on src/pages/Markets/Recommended.tsx:
```diff
@@ -1,18 +1,281 @@
+  selectedChainId,
+  setSelectedChainId,
+}) => {
+  const chainIds = Object.values(KnownChainIds)
```
[preferably blocking] This could use a useMemo for referential stability

gomesalexandre replied:
I like stability f64fb8ed66

woodenfurniture commented on src/assets/translations/en/main.json:
```diff
@@ -2585,7 +2586,25 @@
-    "categories": "Categories",
-    "marketsBody": "Nunc vel efficitur ligula, vel sagittis est. Morbi et sapien a ligula venenatis consequat."
+    "categoriesTabTitle": "Categories",
+    "marketsBody": "Nunc vel efficitur ligula, vel sagittis est. Morbi et sapien a ligula venenatis consequat.",
```
👀

gomesalexandre replied:
Just following the Figma spec haha
<img width="1018" alt="image" src="https://github.com/user-attachments/assets/fa01d58f-ed88-4702-b145-3423465951e3">

woodenfurniture commented on src/assets/translations/en/main.json:
```diff
@@ -2585,7 +2586,25 @@
+      },
+      "trending": {
+        "title": "Trending",
+        "subtitle": "These are top assets that have jumped 10% or more"
```
[suggestion] We could parameterize this e.g `"These are top assets that have jumped %{percentage} or more"`

gomesalexandre replied:
5fa241b41a

woodenfurniture commented on scripts/generateAssetData/generateColorMap.ts:
```diff
@@ -69,7 +69,8 @@ const generateColorMap = async () => {
 
   await fs.promises.writeFile(
-    `./scripts/generateAssetData/colorMap/color-map.json`,
+    // note: cwd is the root of the monorepo, *not* the script directory
+    `src/lib/asset-service/service/color-map.json`,
```
[non-blocking] Weird we use backticks here and linter is happy

woodenfurniture commented on src/lib/asset-service/service/colorMap.ts:
[question] Just to confirm, are we shipping the color map into the client for lazy asset fetching, or what's the intent of moving the color map into the asset service?

gomesalexandre replied:
See https://github.com/shapeshift/web/pull/7736#discussion_r1764619478 for disambiguation - the main idea here was consolidating Portals utils from three places (`portfolioSlice`, `lib/market-service` and `scripts/generateAssetData/utils/portals.ts`).
While in the process of doing so, was experiencing much pain in terms of project references, which was caused by `generatedAssetData.json` and the related color map / generated asset index being outside of the project.

Now that the generation scripts alongside their codegen live in `src/lib`, everything is saner and Portals utils are able to live in one place in web *project* (but not necessarily within the context of web running, i.e scripts)

woodenfurniture reviewed (commented):
🩸Love to see so much red ser. Couple q's but code wise looking 🩸🩸🩸 (good)

gomesalexandre commented on package.json:
```diff
@@ -22,6 +22,7 @@
     "dev:web:linked": "GENERATE_SOURCEMAP=false BROWSER=none yarn dev:web",
     "dev:packages": "yarn tsc --build --watch --preserveWatchOutput tsconfig.packages.json",
     "env": "yarn tsx ./scripts/bootstrap-env.ts",
+    "generate:all": "yarn generate:caip-adapters && yarn run generate:color-map && yarn run generate:asset-data && yarn run generate:thor-asset-map && yarn run generate:thor-longtail-tokens",
```
Will cleanup CI to use this in a follow-up

woodenfurniture reviewed (changes requested):
Code-wise looking good, though found a couple of strange things, unsure which ones are strictly blocking since this is WIP.

# Testing notes

❓ Markets page icon same as pools page

![image](https://github.com/user-attachments/assets/c321e39b-d599-494b-862e-9065e2c0eb7f)

❓ Chain selection affects other rows

https://jam.dev/c/8ea1bf53-10d4-462d-a1e6-953b59d71ec6

❌ Empty chain entry in drop-down leading to empty page

https://jam.dev/c/3a9184a1-cbd3-4b2c-9cd3-8ab66c3391cc

gomesalexandre commented:
> ❓ Markets page icon same as pools page

Replaced with the same as the Assets page to match the spec - 7dfee08d9a :

<img width="766" alt="image" src="https://github.com/user-attachments/assets/d3fe24f6-f314-497e-94d2-009a49792f72">
<img width="322" alt="image" src="https://github.com/user-attachments/assets/ad58088a-ad26-44c1-94c4-e4a0b03dcef6">

> ❓ Chain selection affects other rows

Decoupled selection per row in 78d258f11a, nice catch!

> ❌ Empty chain entry in drop-down leading to empty page

Filtered out Arb nova in 8f835320e8

<img width="391" alt="image" src="https://github.com/user-attachments/assets/c6bdcae3-c87a-44a8-a44e-426fe9222ae9">

Also on a tengent re: two last line items, added a notion of supported ChainIds in 81f9492e0a as well as a narrowing of Portals assets at queryparams time- this will
1. ensure rows don't allow selecting unsupported chains
2. Portals assets (and the next ones, as we fetch them) fire a query that has the selected ChainId (or all) as queryparam, which will ensure we have a full row of results - using the `250` max pageLimit, we *may* have gotten *some* results across all chains, but not guaranteed, and also not guaranteed to fill the row.

https://jam.dev/c/c12a7fed-6b13-4611-8806-2c061bd230b9

woodenfurniture reviewed (approved):
Aww yeah baby now I can look at the top moving base assets while enjoying looking at the trending arbitrum assets 🐐
